Celigo.ServiceManager.NetSuite.REST/src/RestClient.cs

```diff
@@ -0,0 +1,203 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.Extensions.Options;
+
+namespace Celigo.ServiceManager.NetSuite.REST
+{
+    public interface IRestClient
+    {
+        Task<HttpResponseMessage> Get(string account, Uri requestUri, string token, string tokenSecret);
+        Task<HttpResponseMessage> Post<T>(string account, Uri requestUri, string token, string tokenSecret, T content);
+    }
+
+    public class RestClient : IRestClient
+    {
+        private readonly HttpClient _httpClient;
+
+        protected readonly string ConsumerKey;
+        protected readonly string ConsumerSecret;
+        
+        public static JsonSerializerOptions SerializerSettings { get; set; } = new JsonSerializerOptions {
+                                                                             IgnoreNullValues = true,
+                                                                             PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+                                                                         };
+
+        public RestClient(HttpClient httpClient, IOptions<RestClientOptions> options)
+            : this(httpClient, options.Value.ConsumerKey, options.Value.ConsumerSecret) { }
+
+        public RestClient(HttpClient httpClient, string consumerKey, string consumerSecret)
+        {
+            if (string.IsNullOrWhiteSpace(consumerKey))
+                throw new ArgumentNullException(nameof(consumerKey), "Consumer Key must be specified.");
+            if (string.IsNullOrWhiteSpace(consumerSecret))
+                throw new ArgumentNullException(nameof(consumerSecret), "Consumer Secrete must be specified");
+
+            _httpClient = httpClient;
+
+            this.ConsumerKey = consumerKey;
+            this.ConsumerSecret = consumerSecret;
+        }
+
+        private void AddQueryParamsFrom(Uri url, SortedDictionary<string, string> signingParams)
+        {
+            string query = url.Query;
+            if (string.IsNullOrEmpty(query)) return;
+
+            var queryParams = HttpUtility.ParseQueryString(query);
+            foreach (var param in queryParams.AllKeys)
+            {
+                signingParams.Add(param, queryParams[param]);
+            }
+        }
+
+        protected virtual string ComputeNonce()
+        {
+            byte[] data = new byte[20];
+            using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+            {
```

Comment:
Just asking, what's the use of this step ?

Reply:
We are building a cryptogrpahically safe random string.